File: recover/src/recover/state.py
from __future__ import annotations

import time
from dataclasses import dataclass
from threading import Lock


@dataclass
class JobRun:
    last_ran_at_monotonic: float
    last_rows: int
    last_outcome: str
    leader: bool


@dataclass
class JobSnapshot:
    name: str
    last_run_ms_ago: int
    last_rows: int
    last_outcome: str
    leader: bool

    def to_dict(self) -> dict:
        return {
            "name": self.name,
            "last_run_ms_ago": self.last_run_ms_ago,
            "last_rows": self.last_rows,
            "last_outcome": self.last_outcome,
            "leader": self.leader,
        }


class Registry:
    def __init__(self) -> None:
        self._lock = Lock()
        self._jobs: dict[str, JobRun] = {}

    def record(self, name: str, rows: int, outcome: str, leader: bool) -> None:
        with self._lock:
            self._jobs[name] = JobRun(
                last_ran_at_monotonic=time.monotonic(),
                last_rows=rows,
                last_outcome=outcome,
                leader=leader,
            )

    def snapshot(self) -> list[JobSnapshot]:
        now = time.monotonic()
        with self._lock:
            return [
                JobSnapshot(
                    name=name,
                    last_run_ms_ago=int((now - run.last_ran_at_monotonic) * 1000),
                    last_rows=run.last_rows,
                    last_outcome=run.last_outcome,
                    leader=run.leader,
                )
                for name, run in self._jobs.items()
            ]

File: recover/src/state.rs
use std::collections::HashMap;
use std::sync::{Arc, Mutex};
use std::time::Instant;

use serde::Serialize;

#[derive(Clone)]
pub struct Registry {
    inner: Arc<Mutex<HashMap<String, JobRun>>>,
}

#[derive(Clone)]
struct JobRun {
    last_ran_at: Instant,
    last_rows: u64,
    last_outcome: &'static str,
    leader: bool,
}

#[derive(Serialize)]
pub struct JobSnapshot {
    pub name: String,
    pub last_run_ms_ago: u128,
    pub last_rows: u64,
    pub last_outcome: String,
    pub leader: bool,
}

impl Registry {
    pub fn new() -> Self {
        Self {
            inner: Arc::new(Mutex::new(HashMap::new())),
        }
    }

    pub fn record(&self, name: &str, rows: u64, outcome: &'static str, leader: bool) {
        let mut g = self.inner.lock().expect("registry mutex poisoned");
        g.insert(
            name.to_string(),
            JobRun {
                last_ran_at: Instant::now(),
                last_rows: rows,
                last_outcome: outcome,
                leader,
            },
        );
    }

    pub fn snapshot(&self) -> Vec<JobSnapshot> {
        let g = self.inner.lock().expect("registry mutex poisoned");
        g.iter()
            .map(|(name, r)| JobSnapshot {
                name: name.clone(),
                last_run_ms_ago: r.last_ran_at.elapsed().as_millis(),
                last_rows: r.last_rows,
                last_outcome: r.last_outcome.to_string(),
                leader: r.leader,
            })
            .collect()
    }
}

impl Default for Registry {
    fn default() -> Self {
        Self::new()
    }
}
